lib/ClangImporter/ClangImporter.cpp

```diff
@@ -2256,6 +2256,120 @@ ClangModuleUnit *ClangImporter::Implementation::getClangModuleForDecl(
+    const SwiftLookupTable::SingleEntry &target) {
+  bool shouldReset = false;
+  bool alreadyReported = reportedDiagnosticTargets.count(target);
+  if (!alreadyReported && !getEagerImportActive()) {
+    shouldReset = setDiagnosticTarget(target);
+    reportedDiagnosticTargets.insert(target);
+  }
+  return !alreadyReported && !getEagerImportActive() && shouldReset;
+}
+
+bool ClangImporter::Implementation::setDiagnosticTarget(
+    const SwiftLookupTable::SingleEntry &target) {
+  assert(target);
+  // In some situations, the ClangImporter is invoked in a reentrant manner,
+  // resulting in a diagnosticTarget already being set. In such a case, we
+  // keep the outer diagnostic target.
+  if (!diagnosticTarget) {
+    diagnosticTarget = target;
+    return true;
+  }
+  return false;
+}
+
+SwiftLookupTable::SingleEntry
+ClangImporter::Implementation::getDiagnosticTarget() {
+  return diagnosticTarget;
+}
+
+void ClangImporter::Implementation::resetDiagnosticTarget(bool shouldReset) {
+  if (shouldReset) {
+    assert(diagnosticTarget);
+    diagnosticTarget = nullptr;
+  }
+}
+
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const SwiftLookupTable::SingleEntry &targetCandidate) const {
+  return diagnosticTarget == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::NamedDecl *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::NamedDecl *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::ModuleMacro *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::ModuleMacro *>() == targetCandidate;
+}
+bool ClangImporter::Implementation::isDiagnosticTarget(
+    const clang::MacroInfo *targetCandidate) const {
+  return diagnosticTarget.dyn_cast<clang::MacroInfo *>() == targetCandidate;
+}
+
+bool ClangImporter::Implementation::setEagerImportActive() {
+  bool shouldReset = !eagerImportActive;
+  eagerImportActive = true;
+  return shouldReset;
+}
+
+void ClangImporter::Implementation::resetEagerImportActive(bool shouldReset) {
+  if (shouldReset) {
+    eagerImportActive = false;
+  }
+}
+
+bool ClangImporter::Implementation::getEagerImportActive() {
+  return eagerImportActive;
+}
+
+void ClangImporter::Implementation::reportNoteConsumingDiagnosticError(
+    const clang::NamedDecl *target, Diagnostic &&error,
+    const clang::SourceLocation &loc) {
+  if (SwiftContext.LangOpts.EnableExperimentalClangImporterDiagnostics &&
+      isDiagnosticTarget(target)) {
+    diagnose(convertSourceLocation(loc), error);
+    for (auto noteIt = pendingErrorNotes.rbegin();
+         noteIt != pendingErrorNotes.rend(); ++noteIt) {
+      if (!noteIt->diag.getLoc()) {
+        noteIt->diag.setLoc(convertSourceLocation(loc));
+      }
+      if (isDiagnosticTarget(noteIt->target)) {
+        diagnose(noteIt->diag.getLoc(), noteIt->diag);
+      }
+    }
+  }
+  pendingErrorNotes.clear();
```

Comment:
It's not obvious that this function has the side effect of clearing all notes. (Genuine question...) is it possible that a future consumer of this API might want to print all the notes for one target, then all the notes for another target? 

Either way, might be better to let the consumer clear the notes after calling this function. What do you think?

Reply:
Yeah, clearing _all_ the notes really is not strictly necessary and I went back and forth on including this a few times. That being said, with the current system I don't see a reason why the caller would not want to clear all the notes. 

With the current system, I don't think that scenario would occur for the same note. That is, you would lazily import target A and consume the associated notes, then import target B and consume target B related notes. Consuming both target A and target B notes in a single import sort of goes against the system, at least how I see it.